File: backend/app/schemas/lorebook.py
"""Lorebook Pydantic schemas (CR-027)."""

from pydantic import BaseModel, Field, field_validator
from typing import List, Optional, Union
from datetime import datetime
from uuid import UUID


class LorebookEntryCreate(BaseModel):
    """Create lorebook entry request."""
    title: str = Field(..., min_length=1, max_length=200)
    content: str = Field(..., min_length=1, max_length=5000)
    tags: List[str] = Field(default_factory=list, max_length=20)
    priority: int = Field(default=0, ge=0)

    @classmethod
    def validate_tags(cls, v: List[str]) -> List[str]:
        for tag in v:
            if len(tag) > 50:
                raise ValueError(f"Tag '{tag}' exceeds 50 chars")
        return v


class LorebookEntryUpdate(BaseModel):
    """Update lorebook entry request."""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = Field(None, min_length=1, max_length=5000)
    tags: Optional[List[str]] = Field(None, max_length=20)
    priority: Optional[int] = Field(None, ge=0)


class LorebookEntryResponse(BaseModel):
    """Lorebook entry response."""
    id: str
    title: str
    content: str
    tags: List[str]
    priority: int
    status: str
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_validator('id', 'created_by', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        return str(v) if isinstance(v, UUID) else v


class LorebookEntryListItem(BaseModel):
    """Lorebook entry list item (no content)."""
    id: str
    title: str
    tags: List[str]
    priority: int
    status: str
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_validator('id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        return str(v) if isinstance(v, UUID) else v


class LorebookEntryListResponse(BaseModel):
    """Paginated lorebook entry list."""
    items: List[LorebookEntryListItem]
    total: int
    page: int
    page_size: int

File: admin/src/api/lorebook.ts
import { api } from './http';

export interface LorebookEntry {
  id: string;
  title: string;
  content: string;
  tags: string[];
  priority: number;
  status: string;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface LorebookListItem {
  id: string;
  title: string;
  tags: string[];
  priority: number;
  status: string;
  updated_at: string;
}

export interface LorebookListResponse {
  items: LorebookListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface LorebookCreateData {
  title: string;
  content: string;
  tags: string[];
  priority: number;
}

export interface LorebookUpdateData {
  title?: string;
  content?: string;
  tags?: string[];
  priority?: number;
}

export const lorebookApi = {
  list(page = 1, pageSize = 20, tag?: string): Promise<LorebookListResponse> {
    const params = new URLSearchParams({ page: String(page), pageSize: String(pageSize) });
    if (tag) params.append('tag', tag);
    return api.get(`/lorebook?${params.toString()}`);
  },

  get(id: string): Promise<LorebookEntry> {
    return api.get(`/lorebook/${id}`);
  },

  create(data: LorebookCreateData): Promise<LorebookEntry> {
    return api.post('/lorebook', data);
  },

  update(id: string, data: LorebookUpdateData): Promise<LorebookEntry> {
    return api.put(`/lorebook/${id}`, data);
  },

  delete(id: string): Promise<void> {
    return api.delete(`/lorebook/${id}`);
  },
};
